src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs

```diff
@@ -120,8 +125,18 @@ private Task ProcessAuthenticationAsync(bool isAsync = false, CancellationToken
 
         private async Task ProcessAuthenticationWithTelemetryAsync(bool isAsync, CancellationToken cancellationToken)
         {
-            NetSecurityTelemetry.Log.HandshakeStart(IsServer, _sslAuthenticationOptions.TargetHost);
-            long startingTimestamp = Stopwatch.GetTimestamp();
+            long startingTimestamp;
+            if (NetSecurityTelemetry.Log.IsEnabled())
+            {
+                NetSecurityTelemetry.Log.HandshakeStart(IsServer, _sslAuthenticationOptions.TargetHost);
+                startingTimestamp = Stopwatch.GetTimestamp();
+            }
+            else
+            {
+                startingTimestamp = 0;
+            }
+
+            using Activity? activity = s_activitySource.StartActivity(ActivityName, IsServer ? ActivityKind.Server : ActivityKind.Client);
```

Comment:
Actually, I'm still not convinced this should be Client in case of server auth.

> the server auth is initiated by the server

We can view any server operation as being initiated by the client and the server together. Eg. to establish an incoming TCP request from a client socket, the server socket needs to `accept()` it. Even though the first packet is sent by the client, the the server needs to be actively participating in the handshake. I don't see how/why would a TLS handshake fundamentally differ.

Also, according to the [spec](https://opentelemetry.io/docs/specs/otel/trace/api/#spankind) `Client` spans need a `Server` span on the other side, so we either keep TLS `Client`-`Server`, or we make them both `Internal`. I was wondering if the spec has any hints if `Internal` is a right choice, but unfortunately it doesn't.

> Internal spans represent operations which do not cross a process boundary. 

Although it's true for `SslStream` since it's just a utility to execute TLS handshakes over other streams, a literal interpretation of this definition means to me that we could also define HTTP spans `Internal`, since it's the underlying transport layer's job to cross/not cross the process boundaries. One can keep all the data flow in-process if they want.

Reply:
Span kind description is really not so helpful - https://github.com/open-telemetry/opentelemetry-specification/pull/4088

In practice, the `server` kind is used as an indication of external application-level request made against the current application. And the generic server throughput could be calculated as the rate of server spans. When someone renders a service-map, they use server spans to render incoming edges to a node representing a service.

TLS handshake on the server side is a tiny piece of server handling and is not comparable to application-level request processing, you probably never want to see it among incoming requests or as an edge on the service map.

So having additional stages of server call handled as server spans would go against current practice.  
The same argument could be made against CLIENT kind (but everyone abuses it already).

TL;DR: I don't have a good answer, but feel strongly against SERVER. Since there is no guidance/prior art, should we do INTERNAL?

Reply:
ah, just found some prior art in Java instrumentations https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/netty/netty-4.1/javaagent/src/test/groovy/Netty41ConnectionSpanTest.groovy - they use INTERNAL kind for DNS and socket connect. 

Let me check if there was any prior discussion and how they ended up there.

Reply:
> Since there is no guidance/prior art, should we do INTERNAL?

If https://github.com/open-telemetry/opentelemetry-specification/pull/4088 cleans up the definition in a way that would aslo encourage making both connect and accept sides of socket INTERNAL (rather than client-server), it would make sense to me.

Reply:
As these are not individual network requests (DNS could be, but its a couple of steps removed) I think internal is OK.

Reply:
span kinds are defined in such a way that it's hard to say what's client and what's internal. DNS calls don't propagate the context and don't have server spans as children.

we also don't know if there would be any real physical DNS call under that span.

So it's a matter of picking the least controversial option among all other options and sticking with it.